### PR TITLE
bpo-36876: Moved Parser/listnode.c statics to interpreter state.

### PR DESCRIPTION
**Pull request by vsajip**
https://bugs.python.org/issue36876

### CODE REVIEW

File: Include/internal/pycore_pystate.h [hunk 1]
```diff
@@ -132,6 +132,15 @@ struct _is {
...
 
     PyObject *audit_hooks;
+/*
+ * See bpo-36876: miscellaneous ad hoc statics have been moved here.
+ */
+    struct {
+        struct {
+            int level;
+            int atbol;
+        } listnode_data;
```
Comment (ericsnowcurrently) [inline Include/internal/pycore_pystate.h]: "_data" is a little redundant here.

```suggestion
        } listnode;
```

File: Include/internal/pycore_pystate.h [hunk 2]
```diff
@@ -132,6 +132,15 @@ struct _is {
...
+    } parser_data;
```
Comment (ericsnowcurrently) [inline Include/internal/pycore_pystate.h]: Likewise with "_data".

```suggestion
    } parser;
```

File: Include/internal/pycore_pystate.h [hunk 3]
```diff
@@ -132,6 +132,15 @@ struct _is {
     struct _warnings_runtime_state warnings;
 
     PyObject *audit_hooks;
+/*
+ * See bpo-36876: miscellaneous ad hoc statics have been moved here.
+ */
+    struct {
```
Comment (ericsnowcurrently) [inline Include/internal/pycore_pystate.h:138]: If you think there will be any other parser/compiler-related state to move to PyInterpreterState then consider making this struct stand-alone and perhaps even creating pycore_compiler.h (or pycore_parser.h) for it.  That way the code doesn't get too cluttered.

File: Parser/listnode.c [hunk 4]
```diff
@@ -15,19 +16,21 @@ PyNode_ListTree(node *n)
...
-static int level, atbol;
-
 static void
 listnode(FILE *fp, node *n)
 {
-    level = 0;
-    atbol = 1;
+    PyInterpreterState *interp = _PyInterpreterState_GET_UNSAFE();
+
+    interp->parser_data.listnode_data.level = 0;
```
Comment (ericsnowcurrently) [inline Parser/listnode.c]: I was going to point to the advantage of having stand-alone structs for places like this (so there isn't so much going on with each line).  However, I actually kind of like how the line captures the indirection clearly. :)

**Review by ericsnowcurrently — APPROVED** [review]
mostly LGTM

I have left just a few small suggestions to consider.

File: Include/internal/pycore_pystate.h (hunk 1 above)
Comment (vsajip) [reply Include/internal/pycore_pystate.h]: Done.

File: Include/internal/pycore_pystate.h (hunk 2 above)
Comment (vsajip) [reply Include/internal/pycore_pystate.h]: Done.

**Comment by vstinner** [comment]
```
interp = _PyInterpreterState_GET_UNSAFE(); 
```

You can add listnode or parser variable to avoid having to write "interp->parser.listnode". But for that, you need to name the structure in pycore_pystate.h.

**Comment by vsajip** [comment]
I'm not quite sure what you mean. The change was specifically to move those variables into the interpreter state rather than stand-alone variables, as they were previously.

**Comment by vstinner** [comment]
I suggest to add `parser = &interp->parser;` and then use `parser` instead of `interp`. Or something similar with `listnode`.

**Comment by vsajip** [comment]
Ah, OK, I understand now :-)

Not sure what an additional variable buys in this case. I agree, if readability were adversely affected, it would be worth doing. As it is, it seems readable enough.

**Comment by vstinner** [comment]
It was just a minor remark for readability. You're free to ignore it ;-)

**Pull request merged**